compiler/rustc_ty_utils/src/assoc.rs

```diff
@@ -92,6 +93,13 @@ fn impl_item_implementor_ids(tcx: TyCtxt<'_>, impl_id: DefId) -> DefIdMap<DefId>
         .collect()
 }
 
+fn impl_item_trait_item_ids(tcx: TyCtxt<'_>, impl_id: DefId) -> DefIdMap<DefId> {
+    tcx.associated_items(impl_id)
+        .in_definition_order()
+        .filter_map(|item| item.trait_item_def_id.map(|trait_item| (item.def_id, trait_item)))
```

Comment:
All associated items in an impl has the same `trait_item_def_id`, also this is called only for trait impls, so it's always `Some`.

Reply:
We just collect all unsolved impl items firstly, and try to check them after solve other items. One impl item is considered to be used when the self type is used and the corresponding trait item is used. So, getting `trait_item_id` from `impl_item_id` is required to check whether the trait item is used or not.